src/providers/VoiceFocusProvider/index.tsx

```diff
@@ -0,0 +1,167 @@
+  const [voiceFocusTransformer, setVoiceFocusTransformer] = useState<VoiceFocusDeviceTransformer | null>(null);
+
+  // Check whether the value of spec is undefined or empty object
+  const vfSpec = useMemoCompare(spec, (prev: any, next: any) => {
+    const FIXED_EMPTY = {};
+    next === undefined ? FIXED_EMPTY : Object.keys(next).length === 0 ? FIXED_EMPTY : next;
+    return prev && Object.keys(prev).length === 0 && Object.keys(next).length === 0;
+  });
+
+  function useMemoCompare(next: any, compare: any) {
+    const previousRef = useRef();
+    const previous = previousRef.current;
+    const isEqual = compare(previous, next);
+
+    useEffect(() => {
+      if (!isEqual) {
+        previousRef.current = next;
+      }
+    });
+
+    return isEqual ? previous : next;
+  }
+
+  const addVoiceFocus = useCallback(async (device: Device): Promise<Device | VoiceFocusTransformDevice> => {
+    if (!voiceFocusDevice) {
+      return createVoiceFocusDevice(device);
+    }
+    const vf = await voiceFocusDevice.chooseNewInnerDevice(device);
+    setVoiceFocusDevice(vf);
+    return vf;
+  }, [isVoiceFocusSupported]);
+
+  async function createVoiceFocusDevice(inner: Device): Promise<Device | VoiceFocusTransformDevice> {
+    if (!isVoiceFocusSupported) {
+      return inner;
+    }
+    try {
+      const transformer = await getVoiceFocusDeviceTransformer();
+      const device = await transformer?.createTransformDevice(inner);
+      if (device) {
+        setVoiceFocusDevice(device);
+        return device;
+      }
+    } catch (e) {
+      console.warn('Amazon Voice Focus is not supported.', e);
+    }
+    return inner;
+  }
+
+  async function getVoiceFocusDeviceTransformer(): Promise<VoiceFocusDeviceTransformer> {
+    if (voiceFocusTransformer) {
+      return voiceFocusTransformer;
+    }
+
+    const transformer = await VoiceFocusDeviceTransformer.create(spec, options);
+    setVoiceFocusTransformer(transformer);
+    return transformer;
+  }
+
+  let currentPromise: Promise<VoiceFocusDeviceTransformer>;
+
+  useEffect(() => {
+    async function initVoiceFocus() {
+      try {
+        let transformer;
+        const fetch = VoiceFocusDeviceTransformer.create(vfSpec, options);
+
+        fetch.then((value) => {
+          if (fetch === currentPromise) {
+            transformer = value;
+            if (transformer && transformer.isSupported()) {
+              setIsVoiceFocusSupported(true);
+            }
+          } 
+        });
+        currentPromise = fetch;
+        return;
+      } catch (e) {
+        console.warn('Amazon Voice Focus is not supported.', e);
+      }
+      setIsVoiceFocusSupported(false);
+    };
+    initVoiceFocus();
+  }, [vfSpec, options]);
+
+  useEffect(() => {
+    if (isVoiceFocusSupported === true) {
+      console.log('Amazon Voice Focus is supported.');
+    } else if (isVoiceFocusSupported === false) {
+      console.warn('Amazon Voice Focus is not supported.');
+    }
```

Comment:
Consider just

```
if (isVoiceFocusSupported === undefined) {
  return;
}
```

then you can just do a conditional without making people wonder why you're explicitly comparing to true and false.

Reply:
Thanks for the suggestion, updated it